### PR TITLE
Test suite now uses socket pairs instead of memory streams

### PR DESCRIPTION
**Pull request by martinschroeder**
~~This PR contains 2 changes:~~

~~Warnings when using undefined signal constants in a PHPUnit data provider are avoided. Missing constants for signals are defined as needed in the abstract test case.~~

Streams used by test cases are created as connected socket pairs. Most event loop extension do not support arbitrary file descriptors but they all support socket pairs. The created socket pairs use unix domain sockets if they are available. The fallback to INET sockets is needed in order to support Windows.

### CODE REVIEW

**Comment by martinschroeder** [comment]
There is one failing test (testIgnoreRemovedCallback) remaining with lib ev loop. I think that this test is problematic because it assumes that the loop backend triggers events in the same order as writes within the test code.

**Comment by martinschroeder** [comment]
The latest commit fixed the ordering problem with libev. The build failed due to a timing issue with the HHVM build, can't do anything about this...

@WyriHaximus this PR should fix some tests failing due to invalid file descriptors in [reactphp-async-interop-loop
](https://github.com/WyriHaximus/reactphp-async-interop-loop) as well.

**Comment by clue** [comment]
> The build failed due to a timing issue with the HHVM build, can't do anything about this...

I've restarted the HHVM build and this seems to have fixed this for now :+1:

**Comment by martinschroeder** [comment]
Splitting it up sounds good to me.
Should i modify this PR to just address the socket pair stuff and create a new PR for the signal handling later? You could squash to commits during the merge to get rid of the signal stuff.

**Comment by martinschroeder** [comment]
I removed the signal-related part of this PR. Using real sockets in tests enables IO testing using the abstract test class in #25 and #12.

File: tests/AbstractLoopTest.php [hunk 1]
```diff
@@ -16,11 +16,39 @@ public function setUp()
 
     abstract public function createLoop();
 
+    /**
+     * @deprecated Use createSocketPair() instead.
+     */
     public function createStream()
```
Comment (clue) [inline tests/AbstractLoopTest.php]: Should we remove this altogether?

File: tests/AbstractLoopTest.php [hunk 2]
```diff
@@ -16,11 +16,39 @@ public function setUp()
...
+    {
+        return in_array('unix', stream_get_transports(), true);
+    }
 
+    public function createSocketPair()
+    {
+        if ($this->isUnixSocketSupported()) {
+            $sockets = stream_socket_pair(STREAM_PF_UNIX, STREAM_SOCK_STREAM, STREAM_IPPROTO_IP);
+        } else {
+            $sockets = stream_socket_pair(STREAM_PF_INET, STREAM_SOCK_STREAM, STREAM_IPPROTO_IP);
```
Comment (clue) [inline tests/AbstractLoopTest.php]: How can I test this? Shouldn't this use `STREAM_IPPROTO_TCP`? Do we need both branches?

File: tests/AbstractLoopTest.php [hunk 3]
```diff
@@ -16,11 +16,39 @@ public function setUp()
...
+        }
+        
+        foreach ($sockets as $socket) {
+            if (function_exists('stream_set_read_buffer')) {
```
Comment (clue) [inline tests/AbstractLoopTest.php]: This check could be performed before the loop. However, this is lacking some documentation, why is this needed in the first place? And why do we not bother if it doesn't exist?

File: tests/AbstractLoopTest.php [hunk 4]
```diff
@@ -16,11 +16,39 @@ public function setUp()
...
+            }
+        }
+        
+        return $sockets;
+    }
+    
+    /**
+     * @deprecated Use createSocketPair() and fwrite() instead.
+     */
     public function writeToStream($stream, $content)
```
Comment (clue) [inline tests/AbstractLoopTest.php]: Should we remove this altogether?

File: tests/AbstractLoopTest.php [hunk 5]
```diff
@@ -167,71 +195,76 @@ public function emptyRunShouldSimplyReturn()
...
-        );
-
+        $this->loop->addTimer(1, function () {
+            $this->fail('Timer was executed.');
+        });
+        
+        $this->loop->nextTick(function () {
+            $this->loop->stop();
+        });
+
```
Comment (clue) [inline tests/AbstractLoopTest.php]: Unfortunately this PR contains plenty of unrelated formatting changes (see above and below) which make this kind of hard to review. Can you revert these changes so we can focus on what's important here? Thanks!

File: tests/AbstractLoopTest.php (hunk 2 above)
Comment (martinschroeder) [inline tests/AbstractLoopTest.php]: I could turn this into one branch like this:
```php
$domain = $this->isUnixSocketSupported() ? STREAM_PF_UNIX : STREAM_PF_INET;
$sockets = stream_socket_pair($domain, STREAM_SOCK_STREAM, STREAM_IPPROTO_IP);
```
This ensures that it is allways the same function call.

File: tests/AbstractLoopTest.php (hunk 3 above)
Comment (martinschroeder) [inline tests/AbstractLoopTest.php]: Disabling the read buffers is necessary because some event loop extensions will not report streams as readable if PHP does read buffering. Setting the buffer to 0 will ensure that all backends will work on the socket directly.

File: tests/AbstractLoopTest.php (hunk 1 above)
Comment (martinschroeder) [inline tests/AbstractLoopTest.php]: Removing the method would be a better solution. I just marked it as deprecated to keep it backward compatible.

File: tests/AbstractLoopTest.php (hunk 5 above)
Comment (martinschroeder) [inline tests/AbstractLoopTest.php]: I will have to start from scratch to undo the formatting. I will look into it later today or maybe tomorrow.

**Comment by martinschroeder** [comment]
I fixed the code formatting. Unfortunately the HHVM build failed again due to a timing problem...

**Comment by jsor** [comment]
@martinschroeder You should probably rebase/merge the master branch to get b946301bfc0f3996855c5f1f8298c04d2701349c in which should fix the HHVM tests.

**Comment by martinschroeder** [comment]
I merged the master branch, this fixed the HHVM test case. 👍

File: tests/AbstractLoopTest.php (hunk 1 above)
Comment (clue) [inline tests/AbstractLoopTest.php]: Can these be removed now? Afaict they're not needed anymore and BC doesn't really apply to our test suite because it's not part of our public API :+1: (see also below)

**Comment by clue** [comment]
Ping @martinschroeder, what do you think about the above suggestion?

Also, may I ask you to rebase this on the new "0.4" branch so we get this in for the upcoming v0.4.3 release? :shipit:

**Comment by martinschroeder** [comment]
I rebased the PR on the 0.4 branch and removed the obsolete / deprecated methods `createStream()` and `writeToStream()`. I squashed all my commits into a single commit to avoid work-in-progress commits in the history.

**Comment by clue** [comment]
Awesome, much appreciated! :+1:

**Comment by WyriHaximus** [comment]
Looking good, thanks 👍
